### ISSUE TITLE
How to test if all data is loaded from COCO JSON?

### ISSUE THREAD

**Issue by anujdutt9** [openvinotoolkit/training_extensions#551]
Hi,
I am training an object detection model using custom dataset which is in the COCO JSON format. In this I have the positive samples with each image having at-least one annotation and some negative examples as well having bbox annotation of [-1,-1,-1,-1]. Both of these are in the same COCO JSON file. 
How do I test that all of this dataset [positive and negative samples] is being used for model training? 
Thanks.

**Comment by morkovka1337** [comment]
Hi.

If understand your question correctly, you have to split the dataset into training and testing parts which should not overlap.
See [here for details](https://en.wikipedia.org/wiki/Training,_validation,_and_test_sets)

**Comment by anujdutt9** [comment]
Hi @morkovka1337 No. That is already being done. What I am trying to figure out is that if I have a dataset in COCO format for object detection, and in there I have images which have bbox annotations corresponding to them, so like all bboxes with a +ve value correspond to a label > 1 and all bboxes with values -1 correspond to label = 0 i.e. Background (these are my -ve samples added into the dataset to reduce False Positives). Both of these combined form my training dataset
Now how do I confirm that both type of bboxes are used for model training. In COCO dataloader I can see that how the bboxes for labels > 0 are handled, but how are the images with label = 0 are handled and are they used for training or not ?
I am using a SSD model btw.

**Comment by GalyaZalesskaya** [comment]
Hi @anujdutt9.
If you didn't define background class (associated with label=0) in the training config as [here](https://github.com/openvinotoolkit/training_extensions/blob/a47211b3dda2a5f613399cd2cb31f6ed656d5a01/models/object_detection/model_templates/face-detection/face-detection-0200/model.py#L101) the annotation of such classes just being omitted and not being used for training. But if you add a background class in the config, the instances with such class (with label=0) will be handled as well as those with label > 0.

**Comment by anujdutt9** [comment]
Hi @GalyaZalesskaya, thanks for the info. So, does adding our classes and leaving an empty space after that means the background class after all the defined classes like this `classes=('face', )` ? Is my understanding correct?

**Comment by anujdutt9** [comment]
Also for the background images, I have the `bboxes=[-1,-1,-1,-1]` 
Does this work with this framework as background boxes or should it be 0's or empty? Any suggestions? Thanks

**Comment by GalyaZalesskaya** [comment]
@anujdutt9 We don't have a specific method to define background classes and define specific post-processing for such instances. So, my suggestion was to add background class to config like `classes=('foreground', 'background',)`. Leaving a comma and an empty space doesn't have any additional sense here. But it will work only if bbox values contain real boxes with the non-empty area. The logic is the following:
1. All instances with a class that wasn't defined in the training config are being omitted. 
2. Boxes with empty area or with `[-1,-1,-1,-1]` value are being omitted too. 
3. And images without instances (or if all of them were omitted because of reasons above) are being omitted.

Omitted boxes/images aren't being used in the training.
I hope I understood you correctly and answered your question, but if not, do not hesitate to correct me

**Comment by anujdutt9** [comment]
Hi @GalyaZalesskaya Thanks for your explanation. The issue I have is that I am trying to add negative samples to the dataset to reduce the False Positives. 
One way of doing this is to add more images to the background class. Now I don't want to add annotations for background as background could be anything except the object I want to detect. Hence, I was using the `bbox = [-1,-1,-1,-1]` to represent that.
After adding the `background` label, I do see these images being loaded with the dataloader, now what I need to make sure is that these are being used for training as well.
As per your explanation, it looks like that the way I am defining the bboxes for the background class, most likely they will be left out of the training. So, is there a way to include them into the training so as to utilize these negative samples to reduce the number of False Positives?
Thanks

**Comment by GalyaZalesskaya** [comment]
@anujdutt9 I looked through several issues in the upstream and found a way to do it.
The way is to 
1. add  `filter_empty_gt=False,` parameter in the config after `type='CocoDataset',` - that will keep empty images in annotation
2. change [line](https://github.com/openvinotoolkit/mmdetection/blob/951b23a7ecee7fa79caf7f80d71491b7f555a261/mmdet/datasets/pipelines/loading.py#L374) in external/mmdetection/mmdet/datasets/pipelines/loading.py from 
`if results is None or len(results['gt_bboxes']) == 0:` to `if results is None:` - that will allow to load results for empty images

Talking about a way to define bboxes for background class, the developer [here](https://github.com/open-mmlab/mmdetection/issues/2399#issuecomment-685004157) uses no annotation for background images and as far as with changes above we don't filter empty images, it is fine to use his variant.
If you will use negatives please note that it will also require changing loss/assigner parameters. As for me these changes increased the size of the dataset 2 times and made training less converging 
Thank you for your patience and perseverance.

**Comment by anujdutt9** [comment]
Hi @GalyaZalesskaya Thanks a lot for the update and the link. I have made the changes in the code as per your suggestions, but I still don't see an increase in the number of images being used i.e. no change in dataset. Also, my empty bbox looks like this: `{"id": 47300, "image_id": 9713, "bbox": [], "area": 0, "iscrowd": 0, "category_id": 0, "segmentation": []}`, as for the classes, it looks like this: 
```
train=dict(
        type='CocoDataset',
        # Update: include samples with no annotations
        filter_empty_gt=False,
        classes=('head', 'body', 'background'),
        .........
```
What am I missing here?

**Comment by GalyaZalesskaya** [comment]
@anujdutt9 My suggestion is to omit instance info for the background images entirely. This means that you'll have the image info, but the line as you provided will be omitted from the annotation. It will have only instance info about foreground objects with bboxes value we used to (positive values in xywh format).
In addition, based on the information you gave me, that your purpose is to add images without target objects to the training, I think that keeping `background` label in `classes` field lost the sense. Sorry for the misunderstanding.
